dhall/src/Dhall/TypeCheck.hs

```diff
@@ -3140,6 +3155,47 @@ prettyTypeMessage (HandlerInputTypeMismatch expr0 expr1 expr2) =
+    ErrorMessages {..}
+  where
+    short = "Dependently typed handler"
+
+    long =
+        "Explanation: You can ❰merge❱ the alternatives of a union using a record with one\n\
+        \handler per alternative, like this:                                             \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌─────────────────────────────────────────────────────────────────┐         \n\
+        \    │ let union    = < Left : Natural | Right : Bool >.Left 2         │         \n\
+        \    │ let handlers = { Left = Natural/even, Right = λ(x : Bool) → x } │         \n\
+        \    │ in  merge handlers union : Bool                                 │         \n\
+        \    └─────────────────────────────────────────────────────────────────┘         \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \... but the type of the ❰merge❱ may not depend on the union value.              \n\
+        \                                                                                \n\
+        \For example, the following expression is " <> _NOT <> " valid:                  \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌────────────────────────────────────┐                                      \n\
+        \    │ let handlers = { x = None }        │                                      \n\
+        \    │ let union    = < x : Type >.x Bool │                                      \n\
+        \    │ in  merge handlers union           │                                      \n\
+        \    └────────────────────────────────────┘                                      \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \Your handler for the following alternative:                                     \n\
+        \                                                                                \n\
+        \" <> insert label <> "\n\
+        \                                                                                \n\
+        \... has type:                                                                   \n\
+        \                                                                                \n\
+        \" <> insert handlerType <> "\n\
+        \                                                                                \n\
+        \... and its output type depends on the value of the variable:                   \n\
```

Comment:
I suggest also specifically including the output type here, too.  In other words:

```
... and the handler's output type:

↳ …

... references the handler's input type:

↳ …
```